R-package/src/install.libs.R

```diff
@@ -67,36 +106,20 @@ if (!use_precompile) {
   # Check if Windows installation (for gcc vs Visual Studio)
   if (WINDOWS) {
     if (use_mingw) {
+      print("Trying to build with MinGW")
       cmake_cmd <- paste0(cmake_cmd, " -G \"MinGW Makefiles\" ")
       build_cmd <- "mingw32-make.exe _lightgbm"
       system(paste0(cmake_cmd, " ..")) # Must build twice for Windows due sh.exe in Rtools
     } else {
-      local_vs_def <- ""
-      vs_versions <- c(
-        "Visual Studio 16 2019"
-        , "Visual Studio 15 2017"
-        , "Visual Studio 14 2015"
-      )
-      for (vs in vs_versions) {
-        print(paste0("Trying to build with: '", vs, "'"))
-        vs_def <- paste0(" -G \"", vs, "\" -A x64")
-        tmp_cmake_cmd <- paste0(cmake_cmd, vs_def)
-        try_vs <- system(paste0(tmp_cmake_cmd, " .."))
-        if (try_vs == 0L) {
-          local_vs_def <- vs_def
-          print(paste0("Building with '", vs, "' succeeded"))
-          break
-        } else {
-          unlink("./*", recursive = TRUE) # Clean up build directory
-        }
-      }
-      if (try_vs == 1L) {
-        print("Building with Visual Studio failed. Attempted with MinGW")
+      local_vs_def <- .find_vs_version()
+      if (is.null(local_vs_def)) {
+        print("Building with Visual Studio failed. Attempting with MinGW")
```

Comment:
Is there something more "powerful" than `print` which will be able to show this critically important message even in `R CMD check`?

Reply:
Not that I'm aware of! `R CMD check` is only a concern for us developers, not for our users, so I think just `cat`-ing out the `00install.out` file generated by `R CMD CHECK` is ok.

`print()` is enough for people who run `Rscript build_r.R` per our installation docs to see this message.

Reply:
>  so I think just `cat`-ing out the `00install.out` file generated by `R CMD CHECK` is ok.

Yeah, I just wanted to check logs from time to time at our CIs to ensure that we are really using VS.

Reply:
Hmm, just an idea. Maybe it makes sense to disable auto-fallback to MinGW at our CI by env. variable or by simply removing lines from the build script like in building for docs?

https://github.com/microsoft/LightGBM/blob/8c335352c5dc54fb4c223e16a2b6df9448ca9da7/build_r.R#L72
https://github.com/microsoft/LightGBM/blob/8c335352c5dc54fb4c223e16a2b6df9448ca9da7/docs/conf.py#L242

Reply:
@StrikerRUS Ok now that #2936 and #2963 have been merged, we can come back to this!

I think it's a good idea to have something more reliable that us manually looking at the logs to be sure our CI is using the right compiler.

I have a counter-proposal...instead of introducing more complexity into `install.libs.R` (code that our users run), I think we should put that complexity `test_r_package_windows.ps1` (code that only matters to us in CI).

What do you think about this addition I just made in https://github.com/microsoft/LightGBM/pull/2965/commits/096edeaf919986d84fe4e4195c8e29c08103bb87?

```powershell
# Checking that we actually got the expected compiler. The R package has some logic
# to fail back to MinGW if MSVC fails, but for CI builds we need to check that the correct
# compiler ws used.
if ($env:COMPILER -ne "MINGW") {
  $checks = Select-String -Path $INSTALL_LOG_FILE_NAME -Pattern 'Check for working CXX compiler.*MSVC'
} else {
  $checks = Select-String -Path $INSTALL_LOG_FILE_NAME -Pattern 'Check for working CXX compiler.*mingw'
}

if ($checks.Matches.length -ne 0) {
  Write-Output "The wrong compiler was used. Check the build logs."
  Check-Output $False
}
```

Reply:
It seems that this isn't working though...I see that [this build which found g++ from chocolatey](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=5993&view=logs&j=9f61088a-36ee-57e4-0947-03329fd6ee90&t=2528b315-ee14-54c7-8330-fcfcaf0cbb8e) still passed

Reply:
I like your idea! Have you managed to fix compiler checks?

Reply:
Ok yep they are fixed! Here's a commit on my fork where I intentionally made R fail back to MinGW. You can see that installation works correctly but that `chocolatey`'s `g++` gets used, and that the build fails.

![image](https://user-images.githubusercontent.com/7608904/80931004-db202700-8d7c-11ea-9f6e-4796251cef5f.png)

build: https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=404&view=logs&jobId=9f61088a-36ee-57e4-0947-03329fd6ee90&j=9f61088a-36ee-57e4-0947-03329fd6ee90&t=2528b315-ee14-54c7-8330-fcfcaf0cbb8e

commit: https://github.com/jameslamb/LightGBM/pull/21/commits/cf6333c7a300d388c13f9fa5091ff01de328cc49